ipn/localapi/localapi.go

```diff
@@ -247,11 +251,69 @@ func (h *Handler) serveBugReport(w http.ResponseWriter, r *http.Request) {
 	} else {
 		h.logf("user bugreport health: ok")
 	}
-	if defBool(r.FormValue("diagnose"), false) {
+	if defBool(r.URL.Query().Get("diagnose"), false) {
 		h.b.Doctor(r.Context(), logger.WithPrefix(h.logf, "diag: "))
 	}
 	w.Header().Set("Content-Type", "text/plain")
-	fmt.Fprintln(w, logMarker)
+	fmt.Fprintln(w, startMarker)
+
+	// Nothing else to do if we're not in record mode; we wrote the marker
+	// above, so we can just finish our response now.
+	if !defBool(r.URL.Query().Get("record"), false) {
+		return
+	}
+
+	until := time.Now().Add(12 * time.Hour)
+
+	var changed map[string]bool
+	for _, component := range []string{"magicsock"} {
+		if h.b.GetComponentDebugLogging(component).IsZero() {
+			if err := h.b.SetComponentDebugLogging(component, until); err != nil {
+				h.logf("bugreport: error setting component %q logging: %v", component, err)
+				continue
+			}
+
+			mak.Set(&changed, component, true)
+		}
+	}
+	defer func() {
+		for component := range changed {
+			h.b.SetComponentDebugLogging(component, time.Time{})
+		}
+	}()
+
+	// NOTE(andrew): if we have anything else we want to do while recording
+	// a bugreport, we can add it here.
+
+	// Read from the client, or fail when it's canceled.
+	done := make(chan struct{})
+	defer close(done)
+
+	go func() {
+		select {
+		case <-r.Context().Done():
+			r.Body.Close()
+		case <-done:
+		}
+	}()
+	var buf [1]byte
+	_, err := r.Body.Read(buf[:])
+	r.Body.Close()
```

Comment:
Re-reading this, 
Maybe `defer r.Body.Close()` works just as well as this channel

Reply:
Also just realized this is the request body which is automatically closed

https://pkg.go.dev/net/http#Request.Body

Reply:
Just tested this and it does appear to work now that we're just waiting for the client to close the connection, not send a byte; PTAL?

Reply:
So much simpler! Thanks

Reply:
And same to you! I'm much happier with this now TBH.